File: agent-js/src/opentelemetry.ts
import type { ModelUsage } from "@hoangvvo/llm-sdk";
import { context, SpanStatusCode, trace, type Span } from "@opentelemetry/api";
import type { AgentToolResult } from "./tool.ts";
import type { AgentResponse } from "./types.ts";

// Initialie the tracer lazily to allow user to have a chance to configure the global tracer provider
const tracer = trace.getTracer("@hoangvvo/llm-agent");

export class AgentSpan {
  agent_name: string;
  method: "run" | "run_stream";
  usage: ModelUsage | null;
  cost: number | null;

  #span: Span;

  constructor(name: string, method: "run" | "run_stream") {
    this.#span = tracer.startSpan(`llm_agent.${method}`);
    this.agent_name = name;
    this.method = method;
    this.usage = null;
    this.cost = null;
  }

  onResponse(response: AgentResponse) {
    // Aggregate usage and cost from model items within output
    for (const item of response.output) {
      if (item.type === "model") {
        const usage = item.usage;
        const cost = item.cost;
        if (usage) {
          this.usage = this.usage ?? { input_tokens: 0, output_tokens: 0 };
          this.usage.input_tokens += usage.input_tokens;
          this.usage.output_tokens += usage.output_tokens;
        }
        if (typeof cost === "number") {
          this.cost = (this.cost ?? 0) + cost;
        }
      }
    }
  }

  onEnd() {
    this.#span.setAttributes({
      "gen_ai.operation.name": "invoke_agent",
      "gen_ai.agent.name": this.agent_name,
      "gen_ai.model.input_tokens": this.usage?.input_tokens ?? undefined,
      "gen_ai.model.output_tokens": this.usage?.output_tokens ?? undefined,
    });
    this.#span.end();
  }

  onError(error: unknown) {
    this.#span.recordException(error as Error);
    this.#span.setStatus({
      code: SpanStatusCode.ERROR,
      message: String(error),
    });
  }

  withContext<T>(fn: () => T): T {
    return context.with(trace.setSpan(context.active(), this.#span), fn);
  }
}

export function startActiveToolSpan(
  toolCallId: string,
  toolName: string,
  toolDescription: string,
  fn: () => Promise<AgentToolResult>,
) {
  return tracer.startActiveSpan("llm_agent.tool", async (span) => {
    try {
      const res = await fn();
      return res;
    } catch (error) {
      span.recordException(error as Error);
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: String(error),
      });
      throw error;
    } finally {
      // https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/#execute-tool-span
      span.setAttributes({
        "gen_ai.operation.name": "execute_tool",
        "gen_ai.tool.call.id": toolCallId,
        "gen_ai.tool.description": toolDescription,
        "gen_ai.tool.name": toolName,
        "gen_ai.tool.type": "function",
      });
      span.end();
    }
  });
}

File: agent-go/opentelemetry.go
package llmagent

import (
	"context"
	"fmt"

	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"
)

// Initialize the tracer lazily to allow user to have a chance to configure the global tracer provider
var tracer = otel.Tracer("github.com/hoangvvo/llm-agent")

// AgentSpan manages the span for an agent run
type AgentSpan struct {
	agentName string
	method    string
	usage     *llmsdk.ModelUsage
	cost      float64
	span      trace.Span
}

// NewAgentSpan creates a new agent span
func NewAgentSpan(ctx context.Context, agentName string, method string) (*AgentSpan, context.Context) {
	spanName := fmt.Sprintf("llm_agent.%s", method)
	newCtx, span := tracer.Start(ctx, spanName)

	return &AgentSpan{
		agentName: agentName,
		method:    method,
		usage:     nil,
		cost:      0,
		span:      span,
	}, newCtx
}

// OnResponse updates the span with response information
func (s *AgentSpan) OnResponse(response *AgentResponse) {
	for _, item := range response.Output {
		if item.Model != nil {
			if item.Model.Usage != nil {
				if s.usage == nil {
					s.usage = &llmsdk.ModelUsage{}
				}
				s.usage.Add(item.Model.Usage)
			}
			if item.Model.Cost != nil {
				s.cost += *item.Model.Cost
			}
		}
	}
}

// OnEnd ends the span and sets the final attributes
func (s *AgentSpan) OnEnd() {
	attrs := []attribute.KeyValue{
		attribute.String("gen_ai.operation.name", "invoke_agent"),
		attribute.String("gen_ai.agent.name", s.agentName),
	}

	if s.usage != nil {
		attrs = append(attrs, attribute.Int64("gen_ai.model.input_tokens", int64(s.usage.InputTokens)))
		attrs = append(attrs, attribute.Int64("gen_ai.model.output_tokens", int64(s.usage.OutputTokens)))
	}

	if s.cost > 0 {
		attrs = append(attrs, attribute.Float64("llm_agent.cost", s.cost))
	}

	s.span.SetAttributes(attrs...)
	s.span.End()
}

// OnError records an error and ends the span
func (s *AgentSpan) OnError(err error) {
	s.span.RecordError(err)
	s.span.SetStatus(codes.Error, err.Error())
	s.span.End()
}

// startActiveToolSpan creates a span for tool execution
func startActiveToolSpan(
	ctx context.Context,
	toolCallID string,
	toolName string,
	toolDescription string,
	fn func(context.Context) (AgentToolResult, error),
) (AgentToolResult, error) {
	spanCtx, span := tracer.Start(ctx, "llm_agent.tool")
	defer func() {
		// Set attributes following OpenTelemetry semantic conventions
		span.SetAttributes(
			attribute.String("gen_ai.operation.name", "execute_tool"),
			attribute.String("gen_ai.tool.call.id", toolCallID),
			attribute.String("gen_ai.tool.description", toolDescription),
			attribute.String("gen_ai.tool.name", toolName),
			attribute.String("gen_ai.tool.type", "function"),
		)
		span.End()
	}()

	res, err := fn(spanCtx)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return AgentToolResult{}, err
	}

	return res, nil
}
